_drafts/embedded-async-rust.md

```diff
@@ -0,0 +1,681 @@
+            Poll::Ready(())
+        } else {
+            Poll::Pending
+        }
+    }
+}
+```
+
+It's taken a while, but we're finally looking at a Rust future implementation
+for a peripheral! Given how much we've covered so far, this implementation
+should be pretty straightforward. Let's break it down.
+
+```rust
+EXTI_WAKERS[self.pin as usize].register(cx.waker());
+```
+
+This line is the key to how this future is driven to completion. Here we copy
+the waker associated with the future to a static array of wakers that each map
+to a specific interrupt line. We'll come back to this in a moment when we look
+at the actual interrupt handler.
+
+```rust
+let imr = cpu_regs().imr(0).read();
+if !imr.line(self.pin as _) {
+    Poll::Ready(())
+} else {
+    Poll::Pending
+}
+```
+
+Here we check if the interrupt has fired, and return `Poll::Ready` if it has. If
+it hasn't fired yet, we return `Poll::Pending`. This is the pattern we
+established earlier for futures. If there is still work to be done, in the case
+of this future waiting for an interrupt, we return `Poll::Pending`.
+
+```rust
+pub unsafe fn on_irq() {
+    #[cfg(not(any(exti_c0, exti_g0, exti_l5, exti_u5, exti_h5, exti_h50)))]
+    let bits = EXTI.pr(0).read().0;
+    #[cfg(any(exti_c0, exti_g0, exti_l5, exti_u5, exti_h5, exti_h50))]
+    let bits = EXTI.rpr(0).read().0 | EXTI.fpr(0).read().0;
+
+    // Mask all the channels that fired.
+    cpu_regs().imr(0).modify(|w| w.0 &= !bits);
+
+    // Wake the tasks
+    for pin in BitIter(bits) {
+        EXTI_WAKERS[pin as usize].wake();
+    }
+
+    // Clear pending
+    #[cfg(not(any(exti_c0, exti_g0, exti_l5, exti_u5, exti_h5, exti_h50)))]
+    EXTI.pr(0).write_value(Lines(bits));
+    #[cfg(any(exti_c0, exti_g0, exti_l5, exti_u5, exti_h5, exti_h50))]
+    {
+        EXTI.rpr(0).write_value(Lines(bits));
+        EXTI.fpr(0).write_value(Lines(bits));
+    }
+}
+```
+
+The last piece of the puzzle is the interrupt handler. Again we won't cover all
+of the register configurations [^5] here. The important part is right in the
+middle where we call `wake` on the waker associated with the interrupt line that
+fired.
+
+```rust
+for pin in BitIter(bits) {
+    EXTI_WAKERS[pin as usize].wake();
+}
+```
+
+This is the key to how our future is driven to completion. When any interrupt
+registered fires, we'll call the associated waker. Under the hood, this will
+throw our `buttony` task back onto the executor's task queue, and it will be
+polled when it gets its next turn.
+
+## Benefits of Async Rust on Embedded Systems
+
+We now understand how Rust futures work under the hood, and the role an executor
+plays in managing and driving futures to completion. The question remains,
+why would we go through all this trouble? What benefit do we gain over a
+traditional preemptive RTOS or a bare-metal super loop? The answer is I/O!
+
+One of the biggest benefits of async Rust is smaller code size. A traditional
```

Comment:
code size? seems like you mean RAM requirements here? Maybe both are true ;)

Reply:
I sure did, oof